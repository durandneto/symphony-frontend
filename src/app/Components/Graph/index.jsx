import React from 'react'
import PropTypes from 'prop-types'
import Container, { Row, Column } from '../Grid'
import BodyFont from '../Typography/Body'

import {
  GraphContainer,
  GraphColumn,
  GraphValue,
  GraphGap,
  GraphHover
} from './styles'

const Graph = props => {

  const getSlope = arr => {

    if ( arr ) {

      let sumItemArray = arr.map(
        item => {
          return item.data[0] + item.data[1]
        })

      const maxValue = Math.max(...sumItemArray) * 1.2 // mitiply by 1.1 to get a margin top
      // handle array null return 1 to graph
      return maxValue + 1
  } else {
    return 1
  }
}

  // get the max height
  const canvasSize = getSlope(props.data)
  // const slopeHeight = 200 / canvasSize

  //calculate the slope by max sum value
  const slope =  100 / canvasSize

  return (
    <Container loading={props.loading}>
      <Row spaceAround>
        <Column>
          <GraphContainer height={"200px"}>
            {
              props.data.map((item, key) => {
                if ( item.label === props.current.label ) {
                  item = props.current
                }
                return (
                  <GraphColumn key={key} onClick={() => {
                    // avoinding click on the same column
                    if ( item.label !== props.current.label ) {
                      props.onEvent({
                        type: "onClick",
                        origin: "GraphBarColumn",
                        data: item
                      })
                    }
                  }}>
                    <GraphGap lineWidth={props.lineWidth} height={100 - (slope * item.data[1]) - (slope * item.data[0])} />
                    <GraphValue color={(item.label === props.current.label) ? "Teal-100" : "Blue-100"} lineWidth={props.lineWidth} height={slope * item.data[1]} />
                    <GraphValue color={(item.label === props.current.label) ? "Light-Green-A700" : "Indigo-A700"} lineWidth={props.lineWidth} height={slope * item.data[0]} />
                    <GraphHover><BodyFont xs color={"Blue-Grey-300"} label={item.label} /></GraphHover>
                  </GraphColumn>
                )
              })
            }
          </GraphContainer>
        </Column>
      </Row>
    </Container>
  )
}

Graph.defaultProps = {
  onEvent: () => {}
}

Graph.propTypes = {
  data: PropTypes.array.isRequired,
  current: PropTypes.shape({
    label: PropTypes.string.isRequired,
    data: PropTypes.array.isRequired
  }).isRequired,
}

export default Graph
