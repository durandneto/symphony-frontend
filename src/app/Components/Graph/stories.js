import React from 'react'
import { action } from '@storybook/addon-actions'
import Graph from '.'


export default {
  title: 'HIGHLIGHT|Graph/default',
}

const data = [
  {
    label: "1.0.0",
    data: [100,100]
  },
  {
    label: "1.0.1",
    data: [0,0]
  },
  {
    label: "1.0.2",
    data: [0,50]
  }
]

export const Horizontal = () => (
  <Graph
    current={data[1]}
    data={data}
    height={500}
    lineWidth={"15px"}
    onEvent={e => {
      console.log(e)
    }}
  />
)

export const Loading = () => (
  <Graph
    loading
    data={[]}
  />
)
