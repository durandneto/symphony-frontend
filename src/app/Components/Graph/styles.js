import styled from 'styled-components'
import Colors from '../Colors'

export const GraphContainer = styled.div`
  display: flex;
  height: 200px;
  position: relative;

  ${({ height }) => height && `
    height: ${height};
  `}

  ${({ loading }) => loading && `
    opacity: 0.5;
  `}
`

export const GraphColumn = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  align-items: center;
  position: relative;
  padding: 5px;
  transition: transform .2s;
  :hover {
    transform: scale(1.2);
    cursor: pointer;
  }
`
export const GraphHover = styled.div`
  bottom: -40px;
  position: absolute;
  transform: rotate(-90deg);
  white-space: nowrap;
`

export const GraphColumnValue = styled.div`
  background-color: blue;
  transition: height 0.3s ease-out;
  height: 0;
  ${({ height }) => height && `
    height: ${height}%;
  `}

  ${({ lineWidth }) => lineWidth && `
    width: ${lineWidth};
  `}
`

export const GraphGap = styled(GraphColumnValue)`
  background-color: ${Colors["Grey-100"]};

  ${({ height }) => height && height === 100 && `
    height: ${height/2}%;
    background-color: ${Colors["Grey-200"]};
  `}
`

export const GraphValue = styled(GraphColumnValue)`
  ${({ color }) => color && `
    background-color: ${Colors[color]};
  `}
  ${({ color }) => color && `
    background-color: ${Colors[color]};
  `}
`

export default GraphContainer