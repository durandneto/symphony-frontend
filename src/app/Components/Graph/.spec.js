import React from 'react'
import renderer from 'react-test-renderer'
import Graph from '.'

const data = [
  {
    label: "1.0.0",
    data: [100,100]
  },
  {
    label: "1.0.1",
    data: [0,0]
  },
  {
    label: "1.0.2",
    data: [0,50]
  }
]


describe('Graph Component', () => {
  it ('Rendering without crashing', () => {
    const result = renderer.create(
      <Graph
        current={data[1]}
        data={data}
        height={500}
        lineWidth={"15px"}
        onEvent={e => {
          console.log(e)
        }}
      />
    ).toJSON()
    expect(result).toMatchSnapshot()
  })
})
