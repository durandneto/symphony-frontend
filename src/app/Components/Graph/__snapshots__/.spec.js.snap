// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Component Rendering without crashing 1`] = `
<div
  className="sc-AxjAm cbwcBb"
>
  <div
    className="sc-AxjAm sc-AxirZ bxGNId"
  >
    <div
      className="sc-AxjAm sc-AxiKw eXQwvW"
    >
      <div
        className="sc-AxhUy hVCIGX"
        height="200px"
      >
        <div
          className="sc-AxgMl iJHzDk"
          onClick={[Function]}
        >
          <div
            className="sc-Axmtr sc-AxmLO bvIDXk"
            height={17.01244813278008}
          />
          <div
            className="sc-Axmtr sc-fzozJi gmKrhF"
            color="Blue-100"
            height={41.49377593360996}
          />
          <div
            className="sc-Axmtr sc-fzozJi geRLAJ"
            color="Indigo-A700"
            height={41.49377593360996}
          />
          <div
            className="sc-AxheI ghDrgk"
          >
            <span
              className="sc-AxhCb cwjNzG"
              color="Blue-Grey-300"
              label="1.0.0"
            >
              1.0.0
            </span>
          </div>
        </div>
        <div
          className="sc-AxgMl iJHzDk"
          onClick={[Function]}
        >
          <div
            className="sc-Axmtr sc-AxmLO jUcSmd"
            height={100}
          />
          <div
            className="sc-Axmtr sc-fzozJi gqzTyP"
            color="Teal-100"
            height={0}
          />
          <div
            className="sc-Axmtr sc-fzozJi gFHVTf"
            color="Light-Green-A700"
            height={0}
          />
          <div
            className="sc-AxheI ghDrgk"
          >
            <span
              className="sc-AxhCb cwjNzG"
              color="Blue-Grey-300"
              label="1.0.1"
            >
              1.0.1
            </span>
          </div>
        </div>
        <div
          className="sc-AxgMl iJHzDk"
          onClick={[Function]}
        >
          <div
            className="sc-Axmtr sc-AxmLO eTizwB"
            height={79.25311203319502}
          />
          <div
            className="sc-Axmtr sc-fzozJi dlJLgq"
            color="Blue-100"
            height={20.74688796680498}
          />
          <div
            className="sc-Axmtr sc-fzozJi kQOcNn"
            color="Indigo-A700"
            height={0}
          />
          <div
            className="sc-AxheI ghDrgk"
          >
            <span
              className="sc-AxhCb cwjNzG"
              color="Blue-Grey-300"
              label="1.0.2"
            >
              1.0.2
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
